### PR TITLE
LPD-32049 Make SPA style application synchronous based on a configuration flag

### PR DESCRIPTION
**Pull request by izaera**

### CODE REVIEW

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/App.js [hunk 1]
```diff
@@ -185,6 +185,15 @@ class App extends EventEmitter {
...
 		this.popstateScrollTop = 0;
 
+		/**
+		 * Whether to preload CSS files prior to surface flipping so that FOUC
+		 * does not happen.
+		 * @type {!Boolean}
+		 * @default false
+		 * @protected
+		 */
+		this.preloadCSS = !!config?.preloadCSS;
```
Comment (markocikos) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/App.js:195]: Why `!!`? Is `config.preloadCSS` something other than a boolean? If it is, can it be string `'false'`? Because `!!'false'` is `true`.

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/HtmlScreen.js [hunk 2]
```diff
@@ -157,6 +157,32 @@ class HtmlScreen extends RequestScreen {
 		);
 	}
 
+	/**
+	 * @Override
+	 */
+	preloadResources(surfaces) {
+		const tracked = this.virtualQuerySelectorAll_(
+			HtmlScreen.selectors.styles
```
Comment (markocikos) [inline modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/HtmlScreen.js:165]: Any reason for the special handling of `selectors` and `permanentResourcesInDoc`? Why are they not handled as all other properties of the `HtmlScreen` class, and accessed with `this`? This is not related to PR, it's ok to follow existing solution, I'm just curious.

**Review by markocikos — COMMENTED** [review]
@izaera Nice solution! Just few minor questions inline.

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/App.js (hunk 1 above)
Comment (izaera) [reply modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/App.js:195]: It is `undefined` in unit tests (don't you love JavaScript :heart_eyes: ?) Regarding `!!` it's the common idiom to make truthy/falsy things true/false. 

It's true that `'false'` yields true, as you say, but that's expected and works like other languages too (more or less... :roll_eyes:  ).

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/HtmlScreen.js (hunk 2 above)
Comment (izaera) [reply modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/HtmlScreen.js:165]: I have no idea, I copy pasted that code from the function that finally applies the styles. Meaning I made sure I preload all the styles the other function will be using.

Perhaps I should refactor this to centralize that code so that it remains the same forever, for both uses... :thinking:

Comment (izaera) [reply modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/screen/HtmlScreen.js:165]: Couldn't make the refactor :shrug: -> https://github.com/liferay-frontend/liferay-portal/pull/4338/commits/740aef501f24359f417353672011a46e94178224

**Comment by izaera** [comment]
ci:forward:force

File: modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/App.js (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/app/App.js:195]: Ah right, `undefined`. JS :hugs:

**Review by markocikos — APPROVED** [review]
@izaera LGTM! :sparkles:

**Comment by izaera** [comment]
This one is green :heavy_check_mark: and the SF error is unrelated. However, I'm leaving it for @dsanz to review it next week and maybe forward it.

**Comment by dsanz** [comment]
continued in https://github.com/liferay-frontend/liferay-portal/pull/4359
